tests/plugins/clipboard/_helpers/pasting.js

```diff
@@ -3,6 +3,48 @@ getDefaultNotification, createFixtures */
 
 'use strict';
 
+// Mock FileReader.
+( function() {
+	var fileMockBase64 = ';base64,fileMockBase64=',
+		fileMockType,
+		readResultMock;
+
+	function FileReaderMock() {
+		this.listeners = {};
+	}
+
+	// Any MIME type.
+	FileReaderMock.setFileMockType = function( type ) {
+		fileMockType = type;
+	};
+
+	// Result can be: load, abort, error.
+	FileReaderMock.setReadResult = function( readResult ) {
+		readResultMock = readResult;
+		if ( !readResultMock ) {
+			readResultMock = 'load';
+		}
+	};
+
+	FileReaderMock.prototype.addEventListener = function( eventName, callback ) {
+		this.listeners[ eventName ] = callback;
+	};
+
+	FileReaderMock.prototype.readAsDataURL = function() {
+		CKEDITOR.tools.setTimeout( function() {
+			this.result = ( readResultMock == 'load' ? 'data:' + fileMockType + fileMockBase64 : null );
+
+			if ( this.listeners[ readResultMock ] ) {
+				this.listeners[ readResultMock ]();
+			}
+		}, 15, this );
+	};
+
+	/* jshint ignore:start */
+	FileReader = FileReaderMock;
+	/* jshint ignore:end */
+} )();
```

Comment:
This code forces mocking of `FileReader` in all tests that include `pasting.js` helper file. Lots of tests use that file and most of them do not touch `FileReader` at all – so the mock is not needed at all. Additionally it breaks expectations – file with helpers shouldn't have any side effects. There is also an issue connected with the fact that the mock is not restored to its original value after tests.

The mocking logic should be wrapped into a function and that function should be exposed like others in this file. Additionally that function should return the original `FileReader` or a function that would restore `FileReader` upon invoking.

Reply:
The current mocking mechanism probably also causes CI failure, e.g. t[ravis-ci.com/github/ckeditor/ckeditor4/jobs/510358416](https://travis-ci.com/github/ckeditor/ckeditor4/jobs/510358416)